### PR TITLE
Jobs

### PR DESCRIPTION
**Pull request by lucasvanhalst**
## TODO
- [x] jobs demo app

### CODE REVIEW

File: src/solutions/common/bizz/__init__.py [hunk 1]
```diff
@@ -157,7 +157,7 @@ class SolutionModule(Enum):
         FORMS: 'forms',
         PARTICIPATION: 'participation',
         REPORTS: 'oca.reports',
-        JCC_APPOINTMENTS: 'oca.jcc_appointments',
+        JCC_APPOINTMENTS: 'jcc-appointments',
```
Comment (RubenMattan) [inline src/solutions/common/bizz/__init__.py:160]: will this not break all current curstomers?

File: src/cron.yaml [hunk 2]
```diff
@@ -71,3 +71,7 @@ cron:
 - description: Schedule deletion of finished forms
   url: /admin/cron/finish_forms
   schedule: every 1 hours
+
+- description: Notifications for jobs
```
Comment (RubenMattan) [inline src/cron.yaml:75]: lets change to 'Service Notifications for jobs'

File: src/solutions/common/bizz/inbox.py [hunk 3]
```diff
@@ -129,7 +129,7 @@ def transl(key, **params):
         if_email_body_2 = transl('if-email-body-2')
         dashboard_trans = transl('dashboard')
         service_email = sln_settings.login.email() if sln_settings.login else service_user.email()
-        btn = u'<a href="https://rogerth.at?email=%(service_email)s" style="%(button_css)s">%(dashboard)s</a' % {
+        btn = u'<a href="https://rogerth.at?email=%(service_email)s" style="%(button_css)s">%(dashboard)s</a>' % {
```
Comment (RubenMattan) [inline src/solutions/common/bizz/inbox.py:132]: can't we use an onzestadapp domain?

File: src/solutions/common/jobs/bizz.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+class GeoCodeErrorException(HttpBadRequestException):
+    def __init__(self, language):
+        msg = translate(language, SOLUTION_COMMON, 'could_not_resolve_job_address')
+        super(GeoCodeErrorException, self).__init__(msg)
+
+
+class JobOfferNotFoundException(HttpNotFoundException):
+    def __init__(self, job_id):
+        super(JobOfferNotFoundException, self).__init__('Job offer not found', {'id': job_id})
```
Comment (RubenMattan) [inline src/solutions/common/jobs/bizz.py:48]: translate?

File: src/solutions/common/jobs/bizz.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    if not job_offer:
+        raise JobOfferNotFoundException(job_id)
+    return job_offer, stats
+
+
+def create_job_offer(service_user, data):
+    # type: (users.User, EditJobOfferTO) -> Tuple[OcaJobOffer, JobOfferStatistics]
+    job_id = OcaJobOffer.allocate_ids(1, parent=parent_ndb_key(service_user, SOLUTION_COMMON))[0]
+    job_offer = OcaJobOffer(key=OcaJobOffer.create_key(service_user, job_id))
+    language = get_solution_settings(service_user).main_language
```
Comment (RubenMattan) [inline src/solutions/common/jobs/bizz.py:72]: lets do this before allocating an id

Comment (lucasvanhalst) [reply src/solutions/common/jobs/bizz.py:72]: why

File: src/solutions/common/jobs/models.py [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+from rogerthat.dal import parent_ndb_key
+from rogerthat.models import NdbModel
+from rogerthat.models.jobs import JobOfferFunction, JobOfferEmployer, JobOfferLocation, JobOfferContract, \
+    JobOfferContactInformation
+from rogerthat.rpc import users
+from solutions import SOLUTION_COMMON
+
+
+class ContractType(Enum):
+    FULLTIME = 'contract_type_001'
```
Comment (RubenMattan) [inline src/solutions/common/jobs/models.py:30]: link to RT contract types

File: src/solutions/common/bizz/inbox.py (hunk 3 above)
Comment (lucasvanhalst) [reply src/solutions/common/bizz/inbox.py:132]: the whole inbox email should be restyled but let's not do that in this pr

File: src/solutions/common/bizz/__init__.py (hunk 1 above)
Comment (lucasvanhalst) [reply src/solutions/common/bizz/__init__.py:160]: I don't see how changing a translation will break anything

File: src/solutions/common/jobs/models.py [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+
+class JobMatched(NdbModel):
+    source = ndb.IntegerProperty()
+    platform = ndb.TextProperty()
+
+
+class OcaJobOffer(NdbModel):
+    job_domains = ndb.TextProperty(repeated=True)
+    function = ndb.StructuredProperty(JobOfferFunction, indexed=False)  # type: JobOfferFunction
```
Comment (RubenMattan) [inline src/solutions/common/jobs/models.py:77]: use ndb.LocalStructuredProperty(compressed=True) for these

File: src/solutions/common/jobs/models.py (hunk 6 above)
Comment (lucasvanhalst) [reply src/solutions/common/jobs/models.py:30]: what do you mean

File: src/solutions/common/jobs/notifications.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    settings.emails = data.emails
+    settings.put()
+    return settings
+
+
+def send_job_notifications_for_service(jobs_settings_key, min_date, max_date):
+    # type: (ndb.Key, datetime, datetime) -> None
+    jobs_settings = jobs_settings_key.get()  # type: JobsSettings
+    service_user = users.User(jobs_settings_key.parent().id())
+    from dateutil.relativedelta import relativedelta
```
Comment (RubenMattan) [inline src/solutions/common/jobs/notifications.py:70]: lets move import to the top of the funtion

File: src/solutions/common/jobs/notifications.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    min_date = datetime.now() - relativedelta(days=1)
+    solicitations = JobSolicitation.list_unseen_by_service(service_user, min_date, max_date) \
+        .fetch(None)  # type: List[JobSolicitation]
+    if not solicitations:
+        logging.debug('No new updates for jobs from service %s', service_user)
+        return
+    sln_settings = get_solution_settings(jobs_settings.service_user)
+    language = sln_settings.main_language
+
+    def t(key, **kwargs):
```
Comment (RubenMattan) [inline src/solutions/common/jobs/notifications.py:80]: unused

File: src/solutions/common/jobs/solicitations.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+
+    first_message = JobSolicitationMessage(parent=solicitation.key)
+    first_message.message = request.message
+    first_message.reply = False
+    first_message.put()
+
+    solicitation.chat_key = chat_key
+    solicitation.status = JobSolicitationStatus.INITIALIZING
+    solicitation.last_message_key = first_message.key
+    solicitation.put()
```
Comment (RubenMattan) [inline src/solutions/common/jobs/solicitations.py:220]: stats.add_unread(solicitation_key) is missing

File: src/solutions/common/jobs/solicitations.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    logging.debug('Removing chat for job %d, solicitation %d', job_id, solicitation_id)
+    offer = JobOffer.get_by_source(JobOfferSourceType.OCA, str(job_id))  # type: JobOffer
+    keys = [JobMatch.create_key(member.toAppUser(), offer.id),
+            JobSolicitation.create_key(service_user, job_id, solicitation_id)]
+    job_match, solicitation = ndb.get_multi(keys)  # type: JobMatch, JobSolicitation
+    job_match.chat_key = None
+    solicitation.status = JobSolicitationStatus.DISABLED
+    solicitation.chat_key = None
+    solicitation.put()
+    ndb.put_multi([job_match, solicitation])
```
Comment (RubenMattan) [inline src/solutions/common/jobs/solicitations.py:172]: stats.add_read(solicitation_key) is missing

File: src/solutions/common/bizz/__init__.py (hunk 1 above)
Comment (RubenMattan) [reply src/solutions/common/bizz/__init__.py:160]: Thought this was the module :-)

File: src/solutions/common/bizz/inbox.py (hunk 3 above)
Comment (RubenMattan) [reply src/solutions/common/bizz/inbox.py:132]: Oke

File: src/solutions/common/jobs/bizz.py (hunk 5 above)
Comment (RubenMattan) [reply src/solutions/common/jobs/bizz.py:74]: Less things that can fail between allocation and saving

File: src/solutions/common/jobs/models.py (hunk 6 above)
Comment (RubenMattan) [reply src/solutions/common/jobs/models.py:30]: CONTRACT_TYPES in RT but thats a list apparently so its fine

**Comment by RubenMattan** [comment]
@lucasvanhalst we should have a separate group for jobs from demo apps

File: src/solutions/common/jobs/solicitations.py (hunk 11 above)
Comment (lucasvanhalst) [reply src/solutions/common/jobs/solicitations.py:172]: not in this case, they might not have seen the message before the user deleted the chat

File: src/solutions/common/jobs/solicitations.py (hunk 10 above)
Comment (lucasvanhalst) [reply src/solutions/common/jobs/solicitations.py:220]: not needed here, it gets added when the first message is send in the chat which we receive via the new message callback

**Pull request merged**